mono/utils/mono-merp.c

```diff
@@ -0,0 +1,490 @@
+
+	const char *appLoggerName; // App loger name (optional)
+	const char *appLogSessionUUID; // App Session ID (optional but very useful)
+	gboolean isOfficeApplication; // Is office application (1 = True 0 = false). Needs to be 1 to continue. (Required by design)
+	gboolean isObjCException; // Is objectiveC Exception. (optional for crash log generation)
+	size_t memVirt; // Virtual memory (pagination) at the time of crash (optional)
+	size_t memRes; // Physical memory allocated at the time of the crash (optional)
+	const char *treSessionID; // Rules engine session ID (optional)
+	const char *exceptionCodeArg; // Exception code (optional)
+	const char *exceptionAddressArg; // Exception address (optional)
+	gboolean isAppRegisteredWithMAU; // Was this app installed with Microsoft Autoupdate (1 = True 0 = false)
+} MERPStruct;
+
+typedef struct {
+	gboolean enable_merp;
+
+	const char *appBundleID;
+	const char *appSignature; 
+	const char *appVersion;
+	const char *merpGUIPath; 
+	gboolean log;
+} MerpOptions;
+
+MerpOptions config;
+
+static void
+append_merp_arch (GString *output, MerpArch arch)
+{
+	switch (arch) {
+		case MerpArchx86_64:
+			g_string_append_printf (output, "01000007\n");
+			break;
+		case MerpArchx86:
+			g_string_append_printf (output, "00000007\n");
+			break;
+		case MerpArchPPC:
+			g_string_append_printf (output, "00000012\n");
+			break;
+		case MerpArchPPC64:
+			g_string_append_printf (output, "01000012\n");
+			break;
+		default:
+			g_assert_not_reached ();
+	}
+}
+
+static MerpArch
+get_merp_arch () {
+#ifdef TARGET_X86
+	return MerpArchx86;
+#elif defined(TARGET_AMD64)
+	return MerpArchx86_64;
+#elif defined(TARGET_POWERPC)
+	return MerpArchPPC;
+#elif defined(TARGET_POWERPC64)
+	return MerpArchPPC64;
+#else
+	g_assert_not_reached ();
+#endif
+}
+
+static void
+append_merp_exctype (GString *output, MERPExcType exc)
+{
+	switch (exc) {
+		case MERP_EXC_FORCE_QUIT:
+			g_string_append_printf (output, "10000000\n");
+			break;
+		case MERP_EXC_SIGSEGV:
+			g_string_append_printf (output, "20000000\n");
+			break;
+		case MERP_EXC_SIGABRT:
+			g_string_append_printf (output, "30000000\n");
+			break;
+		case MERP_EXC_SIGSYS:
+			g_string_append_printf (output, "40000000\n");
+			break;
+		case MERP_EXC_SIGILL:
+			g_string_append_printf (output, "50000000\n");
+			break;
+		case MERP_EXC_SIGBUS:
+			g_string_append_printf (output, "60000000\n");
+			break;
+		case MERP_EXC_SIGFPE:
+			g_string_append_printf (output, "70000000\n");
+			break;
+		case MERP_EXC_SIGTRAP:
+			g_string_append_printf (output, "80000000\n");
+			break;
+		case MERP_EXC_SIGKILL:
+			g_string_append_printf (output, "90000000\n");
+			break;
+		case MERP_EXC_HANG: 
+			g_string_append_printf (output, "02000000\n");
+			break;
+		case MERP_EXC_NONE:
+			// Exception type is optional
+			g_string_append_printf (output, "\n");
+			break;
+		default:
+			g_assert_not_reached ();
```

Comment:
I'd not assert on error reporting, is there any unknown/general code?

Reply:
Not that I see

Reply:
If we ever hit that line, we've had memory unsafety. This assertion only prevents us from sending a malformed telemetry file.

Reply:
The `exc` value comes from `parse_exception_type` so asserting on its value is correct.